mypyc/irbuild/specialize.py

```diff
@@ -219,6 +230,52 @@ def gen_inner_stmts() -> None:
     return retval
 
 
+@specialize_function('builtins.sum')
+def translate_sum_call(builder: IRBuilder, expr: CallExpr, callee: RefExpr) -> Optional[Value]:
+    # specialized implementation is used if:
+    # - only one or two arguments given (if not, sum() has been given invalid arguments)
+    # - first argument is a Generator (there is no benefit to optimizing the performance of eg.
+    #   sum([1, 2, 3]), so non-Generator Iterables are not handled)
+    if not (len(expr.args) in (1, 2)
+            and expr.arg_kinds[0] == ARG_POS
+            and isinstance(expr.args[0], GeneratorExpr)):
+        return None
+
+    gen_expr = expr.args[0]
+    target_type = builder.node_type(expr)
+
+    # handle 'start' argument, if given
+    start_expr: Optional[Expression] = None
+    if len(expr.args) == 2:
+        # ensure call to sum() was improperly constructed
+        if not expr.arg_kinds[1] in (ARG_POS, ARG_NAMED):
+            return None
+        start_expr = expr.args[1]
+    else:
+        if target_type == int_rprimitive:
+            start_expr = IntExpr(0)
+        elif target_type == float_rprimitive:
+            start_expr = FloatExpr(0.0)
+        else:
+            # IntExpr feels better here, but then if the return value of sum was untypehinted and
+            # the result should be 1, it seems to be True instead, unless we initialize it this
+            # way?
```

Comment:
I agree that this seems like it ought to be an `IntExpr`, and I think that it is actually important for cases like:
```
l: List[Any] = [1, 2, 3]
result = sum(x*x for x in l)
```
That will use object_rprimitive as the type, and we want to produce `14` as the result and not `14.0`.

What was an example of a case where a True was getting produced?

I think that probably we want to wrap the compilation of start_expr in a `builder.coerce` before assigning it to retval.

Reply:
good call. all fixed now